cpp/src/parquet/encoding.cc

```diff
@@ -1188,19 +1190,37 @@ int PlainBooleanDecoder::DecodeArrow(
     typename EncodingTraits<BooleanType>::Accumulator* builder) {
   int values_decoded = num_values - null_count;
   if (ARROW_PREDICT_FALSE(num_values_ < values_decoded)) {
+    // A too large `num_values` was requested.
+    ParquetException::EofException();
+  }
+  if (ARROW_PREDICT_FALSE(!bit_reader_->Advance(values_decoded))) {
     ParquetException::EofException();
   }
 
-  PARQUET_THROW_NOT_OK(builder->Reserve(num_values));
-
-  VisitNullBitmapInline(
-      valid_bits, valid_bits_offset, num_values, null_count,
-      [&]() {
-        bool value;
-        ARROW_IGNORE_EXPR(bit_reader_->GetValue(1, &value));
-        builder->UnsafeAppend(value);
-      },
-      [&]() { builder->UnsafeAppendNull(); });
+  if (null_count == 0) {
+    // FastPath: can copy the data directly
+    PARQUET_THROW_NOT_OK(builder->AppendValues(data_, values_decoded, NULLPTR,
+                                               total_num_values_ - num_values_));
+  } else {
+    int64_t previous_offset = 0;
+    int64_t previous_value_offset = 0;
+    PARQUET_THROW_NOT_OK(::arrow::internal::VisitSetBitRuns(
```

Comment:
If the data is interleaving "null, non-null, null, non-null, ...". Don't know that would performance be

Reply:
I guess the best performance might getting a BitRun scanner here

Reply:
We can start with this. Are there benchmarks for this case?

Reply:
The benchmarks in this PR always have null_probability = 0, right?

Reply:
Yes. Aha previously all null_probability is 0...